roofit/roofit/src/RooLagrangianMorphFunc.cxx

```diff
@@ -0,0 +1,3558 @@
+      if (!hf) {
+        continue;
+      }
+      const RooDataHist &dhist = hf->dataHist();
+      dhist.get(i);
+      RooAbsReal *formula = dynamic_cast<RooAbsReal *>(
+          prod->components().find(Form("w_%s", prod->GetName())));
+      double weight = formula->getVal();
+      unc2 += dhist.weightSquared() * weight * weight;
+      unc += sqrt(dhist.weightSquared()) * weight;
+      val += dhist.weight() * weight;
+    }
+    hist->SetBinContent(i + 1, val);
+    hist->SetBinError(i + 1, correlateErrors ? unc : sqrt(unc2));
+  }
+  /*if (ownResult)
+    delete r;*/
+  return hist;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// count the number of formulas that correspond to the current parameter set
+
+int RooLagrangianMorphFunc::countContributingFormulas() const {
+  int nFormulas = 0;
+  auto mf = std::make_unique<RooRealSumFunc>(*(this->getFunc()));
+  if (!mf)
+    coutE(InputArguments) << "unable to retrieve morphing function"
+                          << std::endl;
+  RooArgSet *args = mf->getComponents();
+  for (auto itr : *args) {
+    RooProduct *prod = dynamic_cast<RooProduct *>(itr);
+    if (prod->getVal() != 0) {
+      nFormulas++;
+    }
+  }
+  return nFormulas;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// check if there is any morphing power provided for the given parameter
+/// morphing power is provided as soon as any two samples provide different,
+/// non-zero values for this parameter
+
+bool RooLagrangianMorphFunc::isParameterUsed(const char *paramname) const {
+  std::string pname(paramname);
+  double val = 0;
+  bool isUsed = false;
+  for (const auto &sample : this->_config.getParamCards()) {
+    double thisval = sample.second.at(pname);
+    if (thisval != val) {
+      if (val != 0)
+        isUsed = true;
+      val = thisval;
+    }
+  }
+  return isUsed;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// check if there is any morphing power provided for the given coupling
+/// morphing power is provided as soon as any two samples provide
+/// different, non-zero values for this coupling
+
+bool RooLagrangianMorphFunc::isCouplingUsed(const char *couplname) const {
+  std::string cname(couplname);
+  const RooArgList *args = this->getCouplingSet();
+  RooAbsReal *coupling = dynamic_cast<RooAbsReal *>(args->find(couplname));
+  if (!coupling)
+    return false;
+  RooLagrangianMorphFunc::ParamSet params = this->getParameters();
+  RooLagrangianMorphFunc *self = const_cast<RooLagrangianMorphFunc *>(this);
+  double val = 0;
+  bool isUsed = false;
+  for (const auto &sample : this->_config.getParamCards()) {
+    self->setParameters(sample.second);
+    double thisval = coupling->getVal();
+    if (thisval != val) {
+      if (val != 0)
+        isUsed = true;
+      val = thisval;
+    }
+  }
+  self->setParameters(params);
```

Comment:
```suggestion
  double val = 0;
  bool isUsed = false;
  for (const auto &sample : this->_config.getParamCards()) {
    this->setParameters(sample.second);
    double thisval = coupling->getVal();
    if (thisval != val) {
      if (val != 0)
        isUsed = true;
      val = thisval;
    }
  }
  this->setParameters(params);
```
Please don't use `const_cast` if not absolutely necessary. It can be very confusing.